pinot-broker/src/main/java/org/apache/pinot/broker/routing/instanceselector/MultiStageStrictReplicaGroupSelector.java

```diff
@@ -0,0 +1,150 @@
+/**
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.apache.pinot.broker.routing.instanceselector;
+
+import com.google.common.annotations.VisibleForTesting;
+import com.google.common.base.Preconditions;
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+import javax.annotation.Nullable;
+import org.apache.helix.model.ExternalView;
+import org.apache.helix.model.IdealState;
+import org.apache.helix.store.zk.ZkHelixPropertyStore;
+import org.apache.helix.zookeeper.datamodel.ZNRecord;
+import org.apache.pinot.broker.routing.adaptiveserverselector.AdaptiveServerSelector;
+import org.apache.pinot.common.assignment.InstancePartitions;
+import org.apache.pinot.common.assignment.InstancePartitionsUtils;
+import org.apache.pinot.common.metrics.BrokerMetrics;
+import org.apache.pinot.spi.config.table.TableType;
+import org.apache.pinot.spi.config.table.assignment.InstancePartitionsType;
+import org.apache.pinot.spi.utils.builder.TableNameBuilder;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+
+/**
+ * Instance selector for multi-stage queries which can ensure that Colocated Tables always leverage Colocated Join
+ * whenever possible. To achieve this, this instance-selector uses InstancePartitions (IP) to determine replica-groups,
+ * as opposed to IdealState used by other instance-selectors. Moreover, this also uses the requestId generated by
+ * Pinot broker to determine the replica-group picked for each table involved in the query, as opposed to using a
+ * member variable. There may be scenarios where an instance in the chosen replica-group is down. In that case, this
+ * strategy will try to pick another replica-group. For realtime tables, this strategy uses only CONSUMING partitions.
```

Comment:
if picking another replica-group is fine, then why do we still need the requestID?

Reply:
`requestId` is the ID created by the Broker for the entire query. We use that to determine which replica-group to pick for all tables. That ensures that the same replica-group is picked for all tables involved in the join.

In context of colocated joins, we will have multiple tables with the same instance partitions. In the happy path, requestId will be used to select the replica-group and we'll select the same set of servers for each table involved.

When an instance in a replica-group is down or a few segments are in a bad state for a table in one of the replicas, we may try to pick another replica-group. There are two possibilities in this case:

1. We still end up picking the same replica-group for all tables involved in the query. This is because if an instance itself is down, it would impact all tables involved, so ideally the next available replica-group should be picked for all tables. This means the query would still use colocated joins.
2. We pick different replica-groups for some of the tables. This would mean that the query won't use colocated join but it would still be able to run.

I think it may also make sense to instead fail the query since technically we weren't able to guarantee that the same replica-group would be used for the full query. I was thinking of following up on this after running some tests in our internal clusters. Hence I marked this feature as Beta for now.

Reply:
Thanks, your explanation is the same as my understanding. My question is if we shall allow the happy path only, given the the strategy name is `strict replica group`. 

Perhaps we can add another strategy or a flag to allow graceful degadation to normal join.

Reply:
Yeah I think that makes sense.

One idea I had was to rename this to "Multistage Replica Group" routing since this allows using a different replica-group if one is unavailable, and add another "Multistage Strict Replica Group" routing which would throw an error if a replica-group is down.

I was thinking of trying it out on our clusters for a few days and see how it goes.

In this PR I can also rename this routing strategy to `multiStageReplicaGroup` (from `multiStageStrictReplicaGroup`).

That way we won't need to make a b/w incompatible change on master in the future. This is marked as Beta anyways.

@yupeng9 lmk your thoughts.

Reply:
Sure. Renaming sounds like a good idea, though I'm curious if we still need a default `MultistageReplicaGroup` without the request ID. Or we always want the request ID for multi-stage query execution?

Reply:
Renamed.

For multi-stage queries any routing strategy would work. This routing strategy is more focused on ensuring we can leverage colocated joins. Perhaps we should rename this to something that explicitly mentions colocated join in the future.